server/conn.go

```diff
@@ -37,50 +37,44 @@ type Conn struct {
 }
 
 var baseConnID uint32 = 10000
+var defaultServer *Server
 
 // NewConn: create connection with default server settings
+//
+// Deprecated: Use [Server.NewConn] instead.
 func NewConn(conn net.Conn, user string, password string, h Handler) (*Conn, error) {
-	p := NewInMemoryProvider()
-	p.AddUser(user, password)
-
-	var packetConn *packet.Conn
-	if defaultServer.tlsConfig != nil {
-		packetConn = packet.NewTLSConn(conn)
-	} else {
-		packetConn = packet.NewConn(conn)
+	if defaultServer == nil {
+		defaultServer = NewDefaultServer()
```

Comment:
Accessing `defaultServer` will cause data race if `NewConn` is called concurrently. I also prefer use a global variable of type `sync.OnceValue` to achieve a) lazily creating the server only when calling it b) caching the computed result for later usage.